src/Adapter/Shipment/DeliveryOptionsProvider.php

```diff
@@ -72,6 +78,44 @@ public function getDeliveryOptions()
         return $parentOutput;
     }
 
+    /**
+     * @return array
+     */
+    public function getProductsByCarrier()
+    {
+        $deliveryOptions = $this->context->cart->getDeliveryOptionList();
+        $currentAddressDeliveryOptions = $deliveryOptions[$this->context->cart->id_address_delivery];
+        $carriers = [];
+
+        foreach ($currentAddressDeliveryOptions as $deliveryOption) {
+            foreach ($deliveryOption['carrier_list'] as $carrier) {
+                $carriers[] = $this->formatCarrierWithProducts($carrier);
+            }
+        }
+
+        return $carriers;
+    }
+
+    public function formatCarrierWithProducts(array $carrierData): array
+    {
+        $carrierProductIds = array_map(function ($product) {
+            return $product['id_product'];
+        }, $carrierData['product_list']);
+
+        $cartProducts = $this->cartPresenter->present($this->context->cart);
```

Comment:
I think it's less performant than doing `$this->context->cart->getProducts()` which should be good enough for what you need, and we avoid requiring additional constructor parameter